aitelemetry/telemetrywrapper.go

```diff
@@ -220,6 +221,55 @@ func (th *telemetryHandle) TrackLog(report Report) {
 	th.client.Track(trace)
 }
 
+// TrackEvent function sends events to appinsights resource. It overrides a few of the existing columns
+// with app information.
+func (th *telemetryHandle) TrackEvent(aiEvent AiEvent) {
+
+	// Initialize new event message
+	event := appinsights.NewEventTelemetry(aiEvent.EventName)
+
+	// OperationId => resourceID (e.g.: NCID)
+	event.Tags.Operation().SetId(aiEvent.ResourceID)
+
+	// Copy the properties, if supplied
+	if aiEvent.Properties != nil {
+		for key, value := range aiEvent.Properties {
+			event.Properties[key] = value
+		}
+	}
+
+	// Acquire read lock to read metadata
+	th.rwmutex.RLock()
+	metadata := th.metadata
+
+	// Add metadata
+	if metadata.SubscriptionID != "" {
+		event.Tags.User().SetAccountId(metadata.SubscriptionID)
+
+		// AnonId => VMName
+		event.Tags.User().SetId(metadata.VMName)
+
+		// SessionId => VMID
+		event.Tags.Session().SetId(metadata.VMID)
+		event.Properties[locationStr] = metadata.Location
+		event.Properties[resourceGroupStr] = metadata.ResourceGroupName
+		event.Properties[vmSizeStr] = metadata.VMSize
+		event.Properties[osVersionStr] = metadata.OSVersion
+		event.Properties[vmIDStr] = metadata.VMID
+		event.Properties[vmNameStr] = metadata.VMName
+	}
+
+	th.rwmutex.RUnlock()
```

Comment:
move runlock before if.. lock needed only for copying metadata to local variable because other thread may update at same time